lib/ecto/changeset.ex

```diff
@@ -1706,55 +1706,68 @@ defmodule Ecto.Changeset do
   """
   def unsafe_validate_unique(changeset, fields, repo, opts \\ []) when is_list(opts) do
     fields = List.wrap(fields)
-    {validations, struct} =
-      case changeset do
-        %Ecto.Changeset{validations: validations, data: %{__struct__: struct}} ->
-          {validations, struct}
-        %Ecto.Changeset{} ->
-          raise ArgumentError, "unsafe_validate_unique/4 does not work with schemaless changesets"
-      end
-    changeset = %{changeset | validations: [{:unsafe_unique, fields} | validations]}
+    changed_fields =
+      changeset.changes
+      |> Map.keys()
+      |> MapSet.new()
 
-    where_clause = for field <- fields do
-      {field, get_field(changeset, field)}
-    end
+    fields_to_validate =
+      fields
+      |> MapSet.new()
 
-    # If we don't have values for all fields, we can't query for uniqueness
-    if Enum.any?(where_clause, &(&1 |> elem(1) |> is_nil())) do
+    if MapSet.disjoint?(changed_fields, fields_to_validate) do
```

Comment:
I think this list is small enough that doing `fields -- Map.keys(changs) == fields` is going to be much more efficient.